# the boston problem
 using regression machine learning model

In [1]:
import numpy as np
import pandas as pd

load dataset


In [5]:

from sklearn.datasets import load_boston
boston_dataset = load_boston()
boston_dataset


{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

data reshaping

In [19]:
boston_df= pd.DataFrame(boston_dataset["data"],columns=boston_dataset["feature_names"])
boston_df["target"]=pd.DataFrame(boston_dataset["target"])
boston_df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  target  
0       15.3  396.90   4.98    24.0  
1       17.8  396.90   9.14    21.6  
2       17.8  392.83   4.03    34.7  
3       18.7  394.63   2.94    33.4  
4       18.7  396.90   5.33    36.2  
..       ...     ...    ...     ...  
501     21.0  391.99   9.67    22.4  
502     21.0  396.90   9.08    20.6  
503     21.0  396.90   5.64    23.9  
504     21.0  393.45   6.48    22.0  
505     21.0  396.90   7.88    11.9  

[506 rows x 14 columns]

data dividing

In [20]:
x= boston_df.drop("target",axis=1)
y= boston_df["target"]

# making regressor model

In [41]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test =train_test_split(x,y,test_size=0.2)

In [48]:

from sklearn.ensemble import RandomForestRegressor
 
np.random.seed(42)

model= RandomForestRegressor().fit(x_train,y_train)

#make a predictions 
y_preds=model.predict(x_test)
y_preds




array([ 8.967, 21.035, 19.836, 11.728, 18.52 , 16.566, 23.363, 33.711,
       20.534, 19.813, 11.186, 15.847, 21.474, 20.097, 13.84 , 11.605,
       25.756, 19.674, 23.253, 20.531, 22.115, 20.939, 20.055, 13.927,
       30.151, 24.254, 15.831, 20.454, 43.289, 11.84 , 33.581, 14.609,
       21.222, 24.472, 20.295, 18.89 , 25.782, 23.493, 15.393, 23.728,
       41.656, 27.064, 20.922, 22.661, 20.718, 24.35 , 31.363, 14.509,
       22.792, 21.405, 31.736,  8.215, 25.87 , 21.386, 28.833, 18.243,
       20.764, 35.135, 46.783, 19.48 , 35.807, 32.538, 23.77 , 21.731,
       20.767, 10.305, 20.268, 25.38 , 22.183, 20.063, 33.417, 15.106,
       14.169, 24.139, 26.218, 47.106, 21.937, 19.532, 23.09 , 14.09 ,
       19.016, 22.205, 20.079, 32.292, 33.997, 22.228, 22.319, 21.054,
       16.034, 19.55 , 48.904, 19.703, 23.679, 18.36 , 18.917, 21.71 ,
        8.737, 20.194, 24.155, 21.166, 34.384, 29.502])

In [44]:
np.array(y_test)

array([ 8.4, 20.3, 20.1,  8.4, 17.6, 15.6, 24.6, 35.4, 16.1, 20.4, 10.9,
       17.2, 22.2, 19.2, 14.9, 12.7, 22. , 16.8, 22.8, 19.9, 26.4, 20. ,
       18.7, 17.8, 30.1, 29.1, 13.8, 19.3, 43.1, 12.8, 35.4, 14. , 15. ,
       23.4, 22.9, 19.7, 25.2, 22.1, 14.2, 24.5, 50. , 26.5, 20.5, 22.9,
       24.5, 24.2, 33.1, 12.7, 23.3, 20.2, 29.6, 10.2, 22.4, 21.2, 30.8,
       17.1, 19.4, 34.7, 43.5, 17.8, 36.4, 33.2, 36.2, 23.2, 22.5,  6.3,
       21.7, 24.4, 21.7, 23.8, 36.1, 11.7, 11.8, 23.7, 23.9, 45.4, 20.3,
       23. , 24.8, 15.2, 20.4, 22.6, 18.9, 34.9, 36.5, 24. , 23.9, 21. ,
       19.4, 19.4, 48.5, 19.5, 23.7, 19.5, 19.6, 20.8,  5. , 20.9, 24.3,
       25. , 33.2, 29.8])

In [46]:
#comparing predictions 

from sklearn.metrics import mean_absolute_error
mean_absolute_error(y_test,y_preds)

1.7417058823529405

In [50]:
model.score(x_test,y_test)

0.9151377110504607

# shows on average our model has just an inaccuracy of about +- 1.7 on guessing target value and has accuracy of 91%